BinaryPropertyListParser improvement, reduced memory allocations

lumley opened the pull request:
Removed most references to copyOfRange inside BinaryPropertyListParser, instead, most NS-Clases now accept the complete byte array and the range that they should copy.

By using this approach, the amount of Byte arrays reserved in memory take about 50-80% less, with a general speed increase of 20-50% depending on the memory stress and the platform used.

*Memory*
![memory](https://cloud.githubusercontent.com/assets/1800418/6927528/52c191ee-d7ee-11e4-8724-0d4ee05ce36b.png)

*Time*
![speed](https://cloud.githubusercontent.com/assets/1800418/6927537/603d5830-d7ee-11e4-97e3-965ac57a938e.png)

Please note the following:
 1. these are just average test results with an over 5MB binary plist parsed that includes NSString, NSNumber, NSArray, NSDictionary and NSDate.
 2. the results include other small adjustments on NSString and NSNumber pending for a pull request. This is my first pull request ever and I mistakenly made all changes to my own master branch, I will continue splitting the changes to individually pull request. Feel free to check some changes to come in my forked master branch.

claunia commented:
Hi @lumley ,

These are generic suggestions, not just applicable to your patch.

1.- Separate comment clean-up or style changes in separate patches, with no functionality changes in them. Some developers like it their style, some projects has a strict coding style. And nonetheless, I've had to navigate thru all the changed comments to see the real code changes.

2.- If you, as an external collaborator, do any code change that breaks the API (as your patch do), you should provide overrides that implement the old method parameters.

Now, particularly to your patch, in almost all integer parsing methods you're passing thru the endindex variable, when, in most of them, it's static (a double is a double, etc). If you use the static cases (10, 8, 4, 2, etc) imho you'll reduce the VM work even more.

lumley commented:
Hi @claunia,

thank you very much for the comments. You are totally right, I was unaware of the API changes that this pull request would introduce, thank you very much for the reminder!

I have reverted the documentation changes and added again the old API headers for compatibility.

About setting the end-index as static. In the specific case you described, the Binary Plist format specifies the offset of each data-type, which is what is calculated during run time in method "readLengthAndOffset", therefore I believe we cannot determine these values before hand.

There may indeed be more cases where we can use static fields and methods, some changes that I have performed in different branches do use static fields and methods, which does effectively increase performance. Regardless of that, the purpose of this pull request was to reduce memory footprint of the class, the increase in speed is just a byproduct of having to perform less reservations.

3breadt commented:
Thank you for that good improvement, also for not including any code style changes and breaking existing APIs.
